### PR TITLE
move ignore parameter support from yaml test client to low level rest client

### PR DESCRIPTION
**Pull request by javanna**
All the language clients support a special `ignore` parameter that doesn't get passed to elasticsearch with the request, but is used to indicate which error code(s) should not lead to an exception if returned for a specific request.

Moving this to the low level REST client will allow the high level REST client to make use of it too, for instance so that it doesn't have to intercept ResponseExceptions when the get api returns a 404, which is in that specific case a perfectly valid response code (e.g. the body returned is not an exception in that case).

### CODE REVIEW

**Comment by javanna** [comment]
@nik9000 can you have a look please?

File: client/rest/src/main/java/org/elasticsearch/client/RestClient.java [hunk 1]
```diff
@@ -282,15 +283,36 @@ public void performRequestAsync(String method, String endpoint, Map<String, Stri
     public void performRequestAsync(String method, String endpoint, Map<String, String> params,
                                     HttpEntity entity, HttpAsyncResponseConsumerFactory httpAsyncResponseConsumerFactory,
                                     ResponseListener responseListener, Header... headers) {
-        URI uri = buildUri(pathPrefix, endpoint, params);
+        Objects.requireNonNull(params, "params must not be null");
+        Map<String, String> requestParams = new HashMap<>(params);
+        Set<Integer> ignoreErrorCodes = new HashSet<>();
```
Comment (s1monw) [inline client/rest/src/main/java/org/elasticsearch/client/RestClient.java]: should we initialize this conditionally and if `ignoreString == null` we just use `Collections.emptySet()`?

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ClientYamlTestClient.java [hunk 2]
```diff
@@ -59,7 +57,7 @@
      * Query params that don't need to be declared in the spec, they are supported by default.
      */
     private static final Set<String> ALWAYS_ACCEPTED_QUERY_STRING_PARAMS = Sets.newHashSet(
-            "error_trace", "filter_path", "human", "pretty", "source");
+            "ignore", "error_trace", "human", "filter_path", "pretty", "source");
```
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ClientYamlTestClient.java]: now that we support this should we document it somewhere?

File: client/rest/src/main/java/org/elasticsearch/client/RestClient.java [hunk 3]
```diff
@@ -282,15 +283,36 @@ public void performRequestAsync(String method, String endpoint, Map<String, Stri
...
+        if (HttpHead.METHOD_NAME.equals(method)) {
+            //404 doesn't cause error if returned for a HEAD request
+            ignoreErrorCodes.add(404);
+        }
+        //ignore is a special parameter supported by the clients, shouldn't be sent to es
+        String ignoreString = requestParams.remove("ignore");
```
Comment (s1monw) [inline client/rest/src/main/java/org/elasticsearch/client/RestClient.java]: does this need javadocs?

**Review by s1monw — CHANGES_REQUESTED** [review]
left some comments

**Comment by javanna** [comment]
@s1monw thanks for the review, I addressed your comments

**Review by s1monw — APPROVED** [review]
LGTM

**Pull request merged**